### DISCUSSION TITLE
Which python package metadata is used by renovate for changelogs, etc?

### DISCUSSION THREAD

**Discussion by wwuck** [renovatebot/renovate#12606] [Suggest an Idea]
I am running renovate on self-hosted Bitbucket Server connecting through a Nexus repository as a proxy to pypi. (I don't think this is relevant to the question, but I can't be 100% sure!)

When I run renovate on python packages, some of the update PRs have the changelog in the PR while others are missing the changelog.  I can see in the docs that renovate somehow fetches some of the changelogs from GitHub releases using GITHUB_COM_TOKEN but it doesn't seem to do this for all packages.

I would like to submit some bug reports or PRs to other python package project repositories so that their release notes appear in renovate, but I can't find where renovate looks in the python package metadata to look up the changelogs source.

One example is `grpcio` and associated `grpcio-*` packages.  The github repository has releases at https://github.com/grpc/grpc/releases but the pypi package information at https://pypi.org/project/grpcio/ only shows the `Homepage` link pointing to https://grpc.io/.  I'm guessing that Google won't change that link to point to their github repository, so there is probably another url in the package metadata that renovate would look up instead?

I'm taking a guess that it might be one or more of [Project-URL](https://packaging.python.org/specifications/core-metadata/#project-url-multiple-use) fields but the Project-URL value is not defined and can contain free-form text so there is no official spec to refer to.

Can we get this metadata usage documented on the renovate docs website so that I can reliably know what to submit in my PRs to problematic python packages?

I am also writing a python package to submit to pypi.org and it would be helpful to know which metadata I should include that renovate is specifically looking for when it scans a python package.

**Comment by wwuck** [comment]
Also, what about packages that have tags for releases but no "GitHub Releases" (eg. https://github.com/python/mypy), or others that have "GitHub Releases" but no changelog information in them (eg. https://github.com/tox-dev/tox/releases)?  Does renovate have the ability to scan and list PR merged on the repository between two releases?

**Reply by viceice** [reply]
https://github.com/renovatebot/renovate/blob/f885e3e4376a830ca7cb3ac8a95a458797ea1852/lib/datasource/pypi/index.ts#L107

There is our pypi logic for parsing metadata.

**Reply by wwuck** [reply]
Thanks
